api/desecapi/pdns.py

```diff
@@ -145,26 +151,399 @@ def get_rrset_datas(domain):
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+        self.transaction = None
+
+    def __enter__(self):
+        self._domain_additions = set()
+        self._domain_deletions = set()
+        self._rr_set_additions = {}
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+        post_save.connect(self._on_rr_post_save, sender=RR)
+        post_delete.connect(self._on_rr_post_delete, sender=RR)
+        post_save.connect(self._on_rr_set_post_save, sender=RRset)
+        post_delete.connect(self._on_rr_set_post_delete, sender=RRset)
+        post_save.connect(self._on_domain_post_save, sender=Domain)
+        post_delete.connect(self._on_domain_post_delete, sender=Domain)
+        # TODO start TPC transaction
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # TODO prepare TPC transaction
+        #  if preparation fails, we don't want to apply pdns changes, hence no try/except
+
+        changes = self._compute_changes()
+        try:
+            self._apply_changes(changes)
+            # TODO commit TPC transaction
+        except (PDNSApplyException, Exception):  # FIXME replace Exception with database-specific exception
+            # at this point, application either the transaction commit or the application of proposed pdns
+            # changes failed for some reason. We're rolling back both. (Rollback will degenerate to no-op
+            # if already rolled back.)
+
+            try:
+                # TODO rollback TPC transaction
+                self._rollback_changes(changes)
+            except (PDNSRollbackException, Exception):  # FIXME replace Exception with database-specific exception
+                # now we're screwed: all changes in the api have been rolled back,
+                # however the rollback on pdns failed
+                # this situation needs to be fixed manually
+                # I.e., at this point, pdns and api are out-of-sync
+                # TODO include as much info as possible
+                raise PDNSOutOfSyncException()
+
+            # pdns was re-synchronized after applied the changes to pdns failed
+            # I.e., pdns and api are in sync, but the planned action failed
+            raise PDNSException()
+
+    def _apply_changes(self, changes, axfr=True):
+        assert all([not c.applied for c in changes])
+
+        axfr_domains = set()
+        for change in changes:
+            change.do()
+            if change.axfr_required():
+                axfr_domains.add(change.domain_name)
+
+        if axfr:
+            self._axfr(axfr_domains)
+
+        return axfr_domains
+
+    def _rollback_changes(self, changes: List[PDNSChange], axfr=True):
+        axfr_domains = set()
+        for change in reversed(changes):
+            if change.applied:
+
+                try:
+                    change.undo()
+                except PDNSException:
+                    raise PDNSRollbackException()
```

Comment:
The other rollbacks should at least be tried, so this need to be put into some list which is evaluated in a `finally` clause or similar

Reply:
The rationale here was that the state will be inconsistent regardless, so why assume one inconsistent state would be better than another. Granted, you could proceed by domain or something, trying to fix each domain's state or so; but the key argument still was that it requires manual fixing anyhow.